### ISSUE TITLE
[FR]: Prefix symbols from Firebase's nanopb dependency

### ISSUE THREAD

**Issue by simonmaddox** [firebase/firebase-ios-sdk#11549]
### Description

1. You require nanopb as part of your Firebase SDK. If we're using nanopb in our own application, the version we're using can conflict with the version you're using. This creates "Duplicate Symbols" errors during linking. This creates a headache when adding or updating nanopb in applications.
2. Currently you leave developers to deal with this issue, where we have to manually prefix our copy of nanopb every time it's updated.
3. Prefix your version of nanopb so that other developers don't need to concern themselves with the inner workings of the Firebase SDK

### Firebase Product(s)

Crashlytics

**Comment by paulb777** [comment]
Thanks for the FR. I'd like to better understand the concern. 

Both CocoaPods and Swift Package Manager are designed to manage versioning to avoid duplication of library dependencies. Both package managers provide tooling to ensure that required dependencies are aligned.

Repercussions of duplicating libraries go beyond duplicate symbols errors, to other concerns like app size bloat and initialization race conditions.

Despite that, I could see an argument for nanopb as a special case since it is small and especially incompatible across its own versions.

**Comment by simonmaddox** [comment]
That’s exactly why I specifically called out nanopb - protobuf generated with one version is not compatible with protobuf generated by another. It’s not just a case of resolving the version that works for both of us - we have to both use the specific version our protobuf was generated for.

We don’t use Cocoapods, but we do use SPM, and are still seeing this issue despite putting our own protobuf and the nanopb source files into it.

For most people using Firebase this FR won’t make a difference, but for those also using nanopb this is a difficult issue to work around.

**Comment by paulb777** [comment]
OK thanks. I made the FR title more specific and actionable.  I believe it still captures the essence of the request.

**Comment by simonmaddox** [comment]
Hi @paulb777. Is there any update on this?

We're concerned that, as part of [Apple's push to increase visibility around the privacy of third party SDKs](https://developer.apple.com/support/third-party-SDK-requirements/), we're going to hit further issues here very soon.

We're expecting that we're going to need to ship nanopb as a signed Swift package (containing an app manifest file) if we want to continue using it, which is going to cause conflicts with Firebase's version of nanopb. 

We're expecting that as part of this we're also going to need to remove our custom symbol prefixing code and fall back to the default nanopb implementation.

**Comment by paulb777** [comment]
@simonmaddox We don't have any near-term plans to prefix our nanopb implementation. We're focusing now on updating our libraries and dependencies for the new privacy and signing requirements.

I don't understand the new concern and why you can't continue to use the custom symbol prefixed version after signing it?

**Comment by simonmaddox** [comment]
> I don't understand the new concern and why you can't continue to use the custom symbol prefixed version after signing it?

Perhaps my understanding is wrong here, but I expected the library would need to be signed by the nanopb author rather than by us.

**Comment by paulb777** [comment]
I cannot speak definitively about Apple's requirements, but consider that nanopb is typically distributed by source, for which signing is not applicable

**Comment by simonmaddox** [comment]
Indeed. I'm not sure how Apple plan to police nanopb use either, but prefixing nanopb is currently a manual (painful) process for us. nanopb is critical to our app, so we're doing what we can to be ready for when Apple makes this required.

It's still very much unknown how this is going to work, which is unfortunate.

**Comment by simonmaddox** [comment]
Hi @paulb777. Now that the privacy and signing updates have shipped, would you consider adding this to your roadmap? It causes us a lot of pain, and I'm sure we're not the only developers using Firebase and nanopb.
